Fix `gson` serialization error of the Refunds object

An error happens when `PATCH /transaction` endpoint is called. 

Initial investigation points to `gson` failed to serialize the `Refunds` interface rather than an object of a concrete class. We will need to deep dive to confirm the root cause.

STACKTRACE: 
```
21:46:15.803 INFO  - o.s.a.p.u.RequestLoggerFilter - "RequestResponseMessage": {"durationMilliseconds": 9, "Request": {"method": "PATCH", "fullPath": "/transactions", "authType": null, "principalName": null, "clientId": "3.134.133.168"}, "Response": {"statusCode": 500, "responseBody": "{"error":"Exception occurred inside setter of org.stellar.anchor.platform.data.JdbcSep31Transaction.refundsJson; nested exception is org.hibernate.PropertyAccessException: Exception occurred inside setter of org.stellar.anchor.platform.data.JdbcSep31Transaction.refundsJson"}"}}
21:46:15.908 ERROR - o.s.a.u.Log - org.springframework.orm.jpa.JpaSystemException: Exception occurred inside setter of org.stellar.anchor.platform.data.JdbcSep31Transaction.refundsJson; nested exception is org.hibernate.PropertyAccessException: Exception occurred inside setter of org.stellar.anchor.platform.data.JdbcSep31Transaction.refundsJson
    at org.springframework.orm.jpa.vendor.HibernateJpaDialect.convertHibernateAccessException(HibernateJpaDialect.java:331)
    at org.springframework.orm.jpa.vendor.HibernateJpaDialect.translateExceptionIfPossible(HibernateJpaDialect.java:233)
    at org.springframework.orm.jpa.AbstractEntityManagerFactoryBean.translateExceptionIfPossible(AbstractEntityManagerFactoryBean.java:551)
    at org.springframework.dao.support.ChainedPersistenceExceptionTranslator.translateExceptionIfPossible(ChainedPersistenceExceptionTranslator.java:61)
    at org.springframework.dao.support.DataAccessUtils.translateIfNecessary(DataAccessUtils.java:242)
    at org.springframework.dao.support.PersistenceExceptionTranslationInterceptor.invoke(PersistenceExceptionTranslationInterceptor.java:152)
    at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
    at org.springframework.data.jpa.repository.support.CrudMethodMetadataPostProcessor$CrudMethodMetadataPopulatingMethodInterceptor.invoke(CrudMethodMetadataPostProcessor.java:174)
    at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
    at org.springframework.aop.interceptor.ExposeInvocationInterceptor.invoke(ExposeInvocationInterceptor.java:97)
    at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
    at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:215)
    at com.sun.proxy.$Proxy163.save(Unknown Source)
    at org.stellar.anchor.platform.data.JdbcSep31TransactionStore.save(JdbcSep31TransactionStore.java:67)
    at org.stellar.anchor.platform.service.TransactionService.patchTransactions(TransactionService.java:108)
    at org.stellar.anchor.platform.controller.PlatformController.patchTransactions(PlatformController.java:40)
    at jdk.internal.reflect.GeneratedMethodAccessor185.invoke(Unknown Source)
    at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.base/java.lang.reflect.Method.invoke(Method.java:566)
    at org.springframework.web.method.support.InvocableHandlerMethod.doInvoke(InvocableHandlerMethod.java:205)
    at org.springframework.web.method.support.InvocableHandlerMethod.invokeForRequest(InvocableHandlerMethod.java:150)
    at org.springframework.web.servlet.mvc.method.annotation.ServletInvocableHandlerMethod.invokeAndHandle(ServletInvocableHandlerMethod.java:117)
    at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.invokeHandlerMethod(RequestMappingHandlerAdapter.java:895)
    at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.handleInternal(RequestMappingHandlerAdapter.java:808)
    at org.springframework.web.servlet.mvc.method.AbstractHandlerMethodAdapter.handle(AbstractHandlerMethodAdapter.java:87)
    at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1067)
    at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:963)
    at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1006)
    at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:880)
    at javax.servlet.http.HttpServlet.service(HttpServlet.java:764)
    at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:227)
    at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)
    at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:53)
    at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189)
    at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)
    at org.stellar.anchor.platform.utils.RequestLoggerFilter.doFilterInternal(RequestLoggerFilter.java:37)
    at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:117)
    at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189)
    at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)
    at org.springframework.web.filter.RequestContextFilter.doFilterInternal(RequestContextFilter.java:100)
    at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:117)
    at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189)
    at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)
    at org.springframework.web.filter.FormContentFilter.doFilterInternal(FormContentFilter.java:93)
    at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:117)
    at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189)
    at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)
    at org.springframework.boot.actuate.metrics.web.servlet.WebMvcMetricsFilter.doFilterInternal(WebMvcMetricsFilter.java:96)
    at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:117)
    at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189)
    at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)
    at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:201)
    at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:117)
    at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189)
    at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)
    at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:197)
    at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:97)
    at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:540)
    at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:135)
    at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:92)
    at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:78)
    at org.apache.catalina.valves.RemoteIpValve.invoke(RemoteIpValve.java:769)
    at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:357)
    at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:382)
    at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:65)
    at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:895)
    at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1732)
    at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49)
    at org.apache.tomcat.util.threads.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1191)
    at org.apache.tomcat.util.threads.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:659)
    at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
    at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: org.hibernate.PropertyAccessException: Exception occurred inside setter of org.stellar.anchor.platform.data.JdbcSep31Transaction.refundsJson
    at org.hibernate.property.access.spi.SetterMethodImpl.set(SetterMethodImpl.java:68)
    at org.hibernate.tuple.entity.AbstractEntityTuplizer.setPropertyValues(AbstractEntityTuplizer.java:681)
    at org.hibernate.tuple.entity.PojoEntityTuplizer.setPropertyValues(PojoEntityTuplizer.java:144)
    at org.hibernate.persister.entity.AbstractEntityPersister.setPropertyValues(AbstractEntityPersister.java:5252)
    at org.hibernate.event.internal.DefaultMergeEventListener.copyValues(DefaultMergeEventListener.java:459)
    at org.hibernate.event.internal.DefaultMergeEventListener.entityIsPersistent(DefaultMergeEventListener.java:205)
    at org.hibernate.event.internal.DefaultMergeEventListener.onMerge(DefaultMergeEventListener.java:178)
    at org.hibernate.event.internal.DefaultMergeEventListener.onMerge(DefaultMergeEventListener.java:70)
    at org.hibernate.event.service.internal.EventListenerGroupImpl.fireEventOnEachListener(EventListenerGroupImpl.java:107)
    at org.hibernate.internal.SessionImpl.fireMerge(SessionImpl.java:833)
    at org.hibernate.internal.SessionImpl.merge(SessionImpl.java:820)
    at jdk.internal.reflect.GeneratedMethodAccessor170.invoke(Unknown Source)
    at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.base/java.lang.reflect.Method.invoke(Method.java:566)
    at org.springframework.orm.jpa.ExtendedEntityManagerCreator$ExtendedEntityManagerInvocationHandler.invoke(ExtendedEntityManagerCreator.java:362)
    at com.sun.proxy.$Proxy152.merge(Unknown Source)
    at jdk.internal.reflect.GeneratedMethodAccessor170.invoke(Unknown Source)
    at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.base/java.lang.reflect.Method.invoke(Method.java:566)
    at org.springframework.orm.jpa.SharedEntityManagerCreator$SharedEntityManagerInvocationHandler.invoke(SharedEntityManagerCreator.java:311)
    at com.sun.proxy.$Proxy152.merge(Unknown Source)
    at org.springframework.data.jpa.repository.support.SimpleJpaRepository.save(SimpleJpaRepository.java:633)
    at jdk.internal.reflect.GeneratedMethodAccessor171.invoke(Unknown Source)
    at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.base/java.lang.reflect.Method.invoke(Method.java:566)
    at org.springframework.data.repository.core.support.RepositoryMethodInvoker$RepositoryFragmentMethodInvoker.lambda$new$0(RepositoryMethodInvoker.java:289)
    at org.springframework.data.repository.core.support.RepositoryMethodInvoker.doInvoke(RepositoryMethodInvoker.java:137)
    at org.springframework.data.repository.core.support.RepositoryMethodInvoker.invoke(RepositoryMethodInvoker.java:121)
    at org.springframework.data.repository.core.support.RepositoryComposition$RepositoryFragments.invoke(RepositoryComposition.java:529)
    at org.springframework.data.repository.core.support.RepositoryComposition.invoke(RepositoryComposition.java:285)
    at org.springframework.data.repository.core.support.RepositoryFactorySupport$ImplementationMethodExecutionInterceptor.invoke(RepositoryFactorySupport.java:639)
    at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
    at org.springframework.data.repository.core.support.QueryExecutorMethodInterceptor.doInvoke(QueryExecutorMethodInterceptor.java:163)
    at org.springframework.data.repository.core.support.QueryExecutorMethodInterceptor.invoke(QueryExecutorMethodInterceptor.java:138)
    at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
    at org.springframework.data.projection.DefaultMethodInvokingMethodInterceptor.invoke(DefaultMethodInvokingMethodInterceptor.java:80)
    at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
    at org.springframework.transaction.interceptor.TransactionInterceptor$1.proceedWithInvocation(TransactionInterceptor.java:123)
    at org.springframework.transaction.interceptor.TransactionAspectSupport.invokeWithinTransaction(TransactionAspectSupport.java:388)
    at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:119)
    at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
    at org.springframework.dao.support.PersistenceExceptionTranslationInterceptor.invoke(PersistenceExceptionTranslationInterceptor.java:137)
    ... 66 more
Caused by: java.lang.reflect.InvocationTargetException
    at jdk.internal.reflect.GeneratedMethodAccessor155.invoke(Unknown Source)
    at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.base/java.lang.reflect.Method.invoke(Method.java:566)
    at org.hibernate.property.access.spi.SetterMethodImpl.set(SetterMethodImpl.java:45)
    ... 107 more
Caused by: java.lang.RuntimeException: Unable to invoke no-args constructor for interface org.stellar.anchor.sep31.Refunds. Registering an InstanceCreator with Gson for this type may fix this problem.
    at com.google.gson.internal.ConstructorConstructor$14.construct(ConstructorConstructor.java:228)
    at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:212)
    at com.google.gson.Gson.fromJson(Gson.java:963)
    at com.google.gson.Gson.fromJson(Gson.java:928)
    at com.google.gson.Gson.fromJson(Gson.java:877)
    at com.google.gson.Gson.fromJson(Gson.java:848)
    at org.stellar.anchor.platform.data.JdbcSep31Transaction.setRefundsJson(JdbcSep31Transaction.java:146)
    ... 111 more
Caused by: java.lang.UnsupportedOperationException: Interface can't be instantiated! Interface name: org.stellar.anchor.sep31.Refunds
    at com.google.gson.internal.UnsafeAllocator.assertInstantiable(UnsafeAllocator.java:117)
    at com.google.gson.internal.UnsafeAllocator$1.newInstance(UnsafeAllocator.java:49)
    at com.google.gson.internal.ConstructorConstructor$14.construct(ConstructorConstructor.java:225)
    ... 117 more
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,5 +1,8 @@
 # Changelog
 
+## 1.2.1
+* Fix gson serialization error of the Refunds object. [#626](https://github.com/stellar/java-stellar-anchor-sdk/issues/626)
+
 ## 1.2.0
 * Add Stellar observer retries with exponential back-off timer [#607](https://github.com/stellar/java-stellar-anchor-sdk/pull/607)
 * Add health check endpoint to the Stellar observer [#602](https://github.com/stellar/java-stellar-anchor-sdk/pull/602)

--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -118,7 +118,7 @@ subprojects {
 
 allprojects {
   group = "org.stellar.anchor-sdk"
-  version = "1.2.0"
+  version = "1.2.1"
 
   tasks.jar {
     manifest {

--- a/core/src/test/kotlin/org/stellar/anchor/sep10/Sep10ServiceTest.kt
+++ b/core/src/test/kotlin/org/stellar/anchor/sep10/Sep10ServiceTest.kt
@@ -5,8 +5,12 @@ import io.mockk.*
 import io.mockk.impl.annotations.MockK
 import java.io.IOException
 import java.security.SecureRandom
+import java.security.cert.X509Certificate
 import java.util.concurrent.TimeUnit
 import java.util.stream.Stream
+import javax.net.ssl.SSLContext
+import javax.net.ssl.TrustManager
+import javax.net.ssl.X509TrustManager
 import okhttp3.OkHttpClient
 import okhttp3.Request
 import okhttp3.mockwebserver.MockResponse
@@ -80,12 +84,8 @@ internal class Sep10ServiceTest {
   private lateinit var sep10Service: Sep10Service
   private val clientKeyPair = KeyPair.random()
   private val clientDomainKeyPair = KeyPair.random()
-  private val httpClient: OkHttpClient =
-    OkHttpClient.Builder()
-      .connectTimeout(10, TimeUnit.MINUTES)
-      .readTimeout(10, TimeUnit.MINUTES)
-      .writeTimeout(10, TimeUnit.MINUTES)
-      .build()
+
+  private lateinit var httpClient: OkHttpClient
 
   @BeforeEach
   fun setUp() {
@@ -108,12 +108,38 @@ internal class Sep10ServiceTest {
 
     this.jwtService = spyk(JwtService(appConfig))
     this.sep10Service = Sep10Service(appConfig, sep10Config, horizon, jwtService)
+
+    this.httpClient = `create httpClient that trust all certificates`()
   }
 
   @AfterEach
   fun tearDown() {
     clearAllMocks()
     unmockkAll()
+  }
+
+  fun `create httpClient that trust all certificates`(): OkHttpClient {
+    val trustAllCerts =
+      arrayOf<TrustManager>(
+        object : X509TrustManager {
+          override fun checkClientTrusted(chain: Array<out X509Certificate>?, authType: String?) {}
+
+          override fun checkServerTrusted(chain: Array<out X509Certificate>?, authType: String?) {}
+
+          override fun getAcceptedIssuers() = arrayOf<X509Certificate>()
+        }
+      )
+
+    // Install the all-trusting trust manager
+    val sslContext = SSLContext.getInstance("SSL")
+    sslContext.init(null, trustAllCerts, SecureRandom())
+    return OkHttpClient.Builder()
+      .connectTimeout(10, TimeUnit.MINUTES)
+      .readTimeout(10, TimeUnit.MINUTES)
+      .writeTimeout(10, TimeUnit.MINUTES)
+      .sslSocketFactory(sslContext.socketFactory, trustAllCerts[0] as X509TrustManager)
+      .hostnameVerifier { _, _ -> true }
+      .build()
   }
 
   @Test

--- a/core/src/test/kotlin/org/stellar/anchor/sep31/Sep31TransactionTest.kt
+++ b/core/src/test/kotlin/org/stellar/anchor/sep31/Sep31TransactionTest.kt
@@ -136,7 +136,7 @@ class Sep31TransactionTest {
   }
 
   @Test
-  fun test_toPlatformApiGetTransactionResponse() {
+  fun `test PlatformApiGetTransactionResponse correctness`() {
     val wantRefunds: Refund =
       Refund.builder()
         .amountRefunded(Amount("90.0000", fiatUSD))
@@ -196,7 +196,7 @@ class Sep31TransactionTest {
   }
 
   @Test
-  fun test_toSep31GetTransactionResponse() {
+  fun `test Sep31GetTransactionResponse correctness`() {
     val refunds =
       Sep31GetTransactionResponse.Refunds.builder()
         .amountRefunded("90.0000")

--- a/platform/src/main/java/org/stellar/anchor/platform/data/JdbcSep31RefundPayment.java
+++ b/platform/src/main/java/org/stellar/anchor/platform/data/JdbcSep31RefundPayment.java
@@ -3,11 +3,9 @@ package org.stellar.anchor.platform.data;
 import lombok.Data;
 import org.stellar.anchor.sep31.RefundPayment;
 
-public class JdbcSep31RefundPayment {
-  @Data
-  public static class JdbcRefundPayment implements RefundPayment {
-    String Id;
-    String amount;
-    String fee;
-  }
+@Data
+public class JdbcSep31RefundPayment implements RefundPayment {
+  String id;
+  String amount;
+  String fee;
 }

--- a/platform/src/main/java/org/stellar/anchor/platform/data/JdbcSep31Refunds.java
+++ b/platform/src/main/java/org/stellar/anchor/platform/data/JdbcSep31Refunds.java
@@ -1,6 +1,7 @@
 package org.stellar.anchor.platform.data;
 
 import com.google.gson.annotations.SerializedName;
+import java.util.ArrayList;
 import java.util.List;
 import lombok.Data;
 import org.stellar.anchor.sep31.RefundPayment;
@@ -15,5 +16,28 @@ public class JdbcSep31Refunds implements Refunds {
   String amountFee;
 
   @SerializedName("payments")
-  List<RefundPayment> refundPayments;
+  List<JdbcSep31RefundPayment> refundPayments;
+
+  @Override
+  public List<RefundPayment> getRefundPayments() {
+    if (refundPayments == null) return null;
+    // getPayments() is made for Gson serialization.
+    List<RefundPayment> payments = new ArrayList<>(refundPayments.size());
+    for (JdbcSep31RefundPayment refundPayment : refundPayments) {
+      payments.add(refundPayment);
+    }
+    return payments;
+  }
+
+  @Override
+  public void setRefundPayments(List<RefundPayment> refundPayments) {
+    this.refundPayments = new ArrayList<>(refundPayments.size());
+    for (RefundPayment rp : refundPayments) {
+      if (rp instanceof JdbcSep31RefundPayment)
+        this.refundPayments.add((JdbcSep31RefundPayment) rp);
+      else
+        throw new ClassCastException(
+            String.format("Error casting %s to JdbcSep31RefundPayment", rp.getClass()));
+    }
+  }
 }

--- a/platform/src/main/java/org/stellar/anchor/platform/data/JdbcSep31Transaction.java
+++ b/platform/src/main/java/org/stellar/anchor/platform/data/JdbcSep31Transaction.java
@@ -143,7 +143,7 @@ public class JdbcSep31Transaction implements Sep31Transaction, SepTransaction {
 
   public void setRefundsJson(String refundsJson) {
     if (refundsJson != null) {
-      this.refunds = gson.fromJson(refundsJson, Refunds.class);
+      this.refunds = gson.fromJson(refundsJson, JdbcSep31Refunds.class);
     }
   }
 

--- a/platform/src/main/java/org/stellar/anchor/platform/data/JdbcSep31TransactionStore.java
+++ b/platform/src/main/java/org/stellar/anchor/platform/data/JdbcSep31TransactionStore.java
@@ -31,7 +31,7 @@ public class JdbcSep31TransactionStore implements Sep31TransactionStore {
 
   @Override
   public RefundPayment newRefundPayment() {
-    return new JdbcSep31RefundPayment.JdbcRefundPayment();
+    return new JdbcSep31RefundPayment();
   }
 
   @Override

--- a/platform/src/test/kotlin/org/stellar/anchor/platform/data/JdbcSep31TransactionTest.kt
+++ b/platform/src/test/kotlin/org/stellar/anchor/platform/data/JdbcSep31TransactionTest.kt
@@ -1,0 +1,61 @@
+package org.stellar.anchor.platform.data
+
+import org.junit.jupiter.api.Assertions.assertEquals
+import org.junit.jupiter.api.Assertions.assertNull
+import org.junit.jupiter.api.Test
+import org.skyscreamer.jsonassert.JSONAssert
+
+class JdbcSep31TransactionTest {
+  private val refundsJsonNoRefundPayment =
+    """
+            {
+              "amount_refunded": "10",
+              "amount_fee": "5",
+              "refundPayments": null
+            }
+        """.trimIndent()
+
+  private val refundsJsonWithRefundPayment =
+    """
+            {
+              "amount_refunded": "10",
+              "amount_fee": "5",
+              "payments": [
+                {
+                  "id": "1",
+                  "amount": "5",
+                  "fee": "1"
+                },
+                {
+                  "id": "2",
+                  "amount": "5",
+                  "fee": "4"
+                }
+              ]
+            }
+        """.trimIndent()
+
+  @Test
+  fun `test JdbcSep31Transaction refunds Json conversion`() {
+    val txn = JdbcSep31Transaction()
+    txn.refundsJson = refundsJsonNoRefundPayment
+    // strict is set to false because refundPayments is omitted in txn.refundsJson when it is set to
+    // null.
+    JSONAssert.assertEquals(txn.refundsJson, refundsJsonNoRefundPayment, false)
+    assertEquals("10", txn.refunds.amountRefunded)
+    assertEquals("5", txn.refunds.amountFee)
+    assertNull(txn.refunds.refundPayments)
+
+    txn.refundsJson = refundsJsonWithRefundPayment
+    JSONAssert.assertEquals(txn.refundsJson, refundsJsonWithRefundPayment, true)
+    assertEquals("10", txn.refunds.amountRefunded)
+    assertEquals("5", txn.refunds.amountFee)
+    assertEquals(2, txn.refunds.refundPayments.size)
+    assertEquals("1", txn.refunds.refundPayments[0].id)
+    assertEquals("5", txn.refunds.refundPayments[0].amount)
+    assertEquals("1", txn.refunds.refundPayments[0].fee)
+    assertEquals("2", txn.refunds.refundPayments[1].id)
+    assertEquals("5", txn.refunds.refundPayments[1].amount)
+    assertEquals("4", txn.refunds.refundPayments[1].fee)
+  }
+}

--- a/platform/src/test/kotlin/org/stellar/anchor/platform/service/TransactionServiceTest.kt
+++ b/platform/src/test/kotlin/org/stellar/anchor/platform/service/TransactionServiceTest.kt
@@ -3,8 +3,11 @@ package org.stellar.anchor.platform.service
 import io.mockk.*
 import io.mockk.impl.annotations.MockK
 import java.time.Instant
-import org.junit.jupiter.api.*
+import org.junit.jupiter.api.AfterEach
 import org.junit.jupiter.api.Assertions.*
+import org.junit.jupiter.api.BeforeEach
+import org.junit.jupiter.api.Test
+import org.junit.jupiter.api.assertThrows
 import org.junit.jupiter.params.ParameterizedTest
 import org.junit.jupiter.params.provider.EnumSource
 import org.stellar.anchor.api.exception.AnchorException
@@ -19,7 +22,7 @@ import org.stellar.anchor.api.shared.RefundPayment
 import org.stellar.anchor.asset.AssetService
 import org.stellar.anchor.asset.ResourceJsonAssetService
 import org.stellar.anchor.event.models.TransactionEvent
-import org.stellar.anchor.platform.data.JdbcSep31RefundPayment.JdbcRefundPayment
+import org.stellar.anchor.platform.data.JdbcSep31RefundPayment
 import org.stellar.anchor.platform.data.JdbcSep31Refunds
 import org.stellar.anchor.platform.data.JdbcSep31Transaction
 import org.stellar.anchor.sep31.*
@@ -78,7 +81,7 @@ class TransactionServiceTest {
     // Mock the store
     every { sep31TransactionStore.newTransaction() } returns JdbcSep31Transaction()
     every { sep31TransactionStore.newRefunds() } returns JdbcSep31Refunds()
-    every { sep31TransactionStore.newRefundPayment() } answers { JdbcRefundPayment() }
+    every { sep31TransactionStore.newRefundPayment() } answers { JdbcSep31RefundPayment() }
 
     // mock time
     val mockStartedAt = Instant.now().minusSeconds(180)

